// eslint-plugin-readable-tailwind/lib/cjs/options/callees/tv.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TV = exports.TV_COMPOUND_VARIANTS_CLASS = exports.TV_VARIANT_VALUES = exports.TV_STRINGS = void 0;
const readable_tailwind_types_rule_js_1 = require("../../types/rule.js");
exports.TV_STRINGS = [
    "tv",
    [
        {
            match: readable_tailwind_types_rule_js_1.MatcherType.String
        }
    ]
];
exports.TV_VARIANT_VALUES = [
    "tv",
    [
        {
            match: readable_tailwind_types_rule_js_1.MatcherType.ObjectValue,
            pathPattern: "^variants.*$"
        }
    ]
];
exports.TV_COMPOUND_VARIANTS_CLASS = [
    "tv",
    [
        {
            match: readable_tailwind_types_rule_js_1.MatcherType.ObjectValue,
            pathPattern: "^compoundVariants\\[\\d+\\]\\.(?:className|class)$"
        }
    ]
];
/** @see https://github.com/nextui-org/tailwind-variants?tab=readme-ov-file */
exports.TV = [
    exports.TV_STRINGS,
    exports.TV_VARIANT_VALUES,
    exports.TV_COMPOUND_VARIANTS_CLASS
];
