// eslint-plugin-readable-tailwind/lib/cjs/tailwind/utils/resolvers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsonResolver = exports.cssResolver = exports.cjsResolver = exports.esmResolver = void 0;
const node_fs_1 = __importDefault(require("node:fs"));
const enhanced_resolve_1 = __importDefault(require("enhanced-resolve"));
const fileSystem = new enhanced_resolve_1.default.CachedInputFileSystem(node_fs_1.default, 30000);
exports.esmResolver = enhanced_resolve_1.default.ResolverFactory.createResolver({
    conditionNames: ["node", "import"],
    extensions: [".mjs", ".js"],
    fileSystem,
    mainFields: ["module"],
    useSyncFileSystemCalls: true
});
exports.cjsResolver = enhanced_resolve_1.default.ResolverFactory.createResolver({
    conditionNames: ["node", "require"],
    extensions: [".js", ".cjs"],
    fileSystem,
    mainFields: ["main"],
    useSyncFileSystemCalls: true
});
exports.cssResolver = enhanced_resolve_1.default.ResolverFactory.createResolver({
    conditionNames: ["style"],
    extensions: [".css"],
    fileSystem,
    mainFields: ["style"],
    useSyncFileSystemCalls: true
});
exports.jsonResolver = enhanced_resolve_1.default.ResolverFactory.createResolver({
    conditionNames: ["json"],
    extensions: [".json"],
    fileSystem,
    useSyncFileSystemCalls: true
});
