// eslint-plugin-readable-tailwind/lib/cjs/rules/no-restricted-classes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.noRestrictedClasses = void 0;
exports.getOptions = getOptions;
const better_tailwindcss_options_default_options_js_1 = require("../options/default-options.js");
const better_tailwindcss_options_descriptions_js_1 = require("../options/descriptions.js");
const better_tailwindcss_utils_rule_js_1 = require("../utils/rule.js");
const better_tailwindcss_utils_utils_js_1 = require("../utils/utils.js");
const defaultOptions = {
    attributes: better_tailwindcss_options_default_options_js_1.DEFAULT_ATTRIBUTE_NAMES,
    callees: better_tailwindcss_options_default_options_js_1.DEFAULT_CALLEE_NAMES,
    restrict: [],
    tags: better_tailwindcss_options_default_options_js_1.DEFAULT_TAG_NAMES,
    variables: better_tailwindcss_options_default_options_js_1.DEFAULT_VARIABLE_NAMES
};
const DOCUMENTATION_URL = "https://github.com/schoero/eslint-plugin-better-tailwindcss/blob/main/docs/rules/no-restricted-classes.md";
exports.noRestrictedClasses = {
    name: "no-restricted-classes",
    rule: {
        create: ctx => (0, better_tailwindcss_utils_rule_js_1.createRuleListener)(ctx, getOptions(ctx), lintLiterals),
        meta: {
            docs: {
                description: "Disallow restricted classes.",
                recommended: false,
                url: DOCUMENTATION_URL
            },
            schema: [
                {
                    additionalProperties: false,
                    properties: {
                        ...better_tailwindcss_options_descriptions_js_1.CALLEE_SCHEMA,
                        ...better_tailwindcss_options_descriptions_js_1.ATTRIBUTE_SCHEMA,
                        ...better_tailwindcss_options_descriptions_js_1.VARIABLE_SCHEMA,
                        ...better_tailwindcss_options_descriptions_js_1.TAG_SCHEMA,
                        restrict: {
                            items: {
                                type: "string"
                            },
                            type: "array"
                        }
                    },
                    type: "object"
                }
            ],
            type: "problem"
        }
    }
};
function lintLiterals(ctx, literals) {
    const { restrict: restrictions } = getOptions(ctx);
    for (const literal of literals) {
        const classes = literal.content;
        const classNames = (0, better_tailwindcss_utils_utils_js_1.splitClasses)(classes);
        const restrict = classNames.filter(className => {
            return restrictions.some(restriction => className.match(restriction));
        });
        for (const restrictedClass of restrict) {
            ctx.report({
                data: {
                    restrictedClass
                },
                loc: (0, better_tailwindcss_utils_utils_js_1.getExactClassLocation)(literal, restrictedClass),
                message: "Restricted class: \"{{ restrictedClass }}\"."
            });
        }
    }
}
function getOptions(ctx) {
    const options = ctx.options[0] ?? {};
    const common = (0, better_tailwindcss_utils_utils_js_1.getCommonOptions)(ctx);
    const restrict = options.restrict ?? defaultOptions.restrict;
    return {
        ...common,
        restrict
    };
}
